warehouse/manage/views/__init__.py

```diff
@@ -1421,6 +1432,118 @@ def add_google_oidc_publisher(self):
+                "warehouse.oidc.add_publisher.ratelimited",
+                tags=["publisher:ActiveState"],
+            )
+            return HTTPTooManyRequests(
+                self.request._(
+                    "There have been too many attempted trusted publisher "
+                    "registrations. Try again later."
+                ),
+                retry_after=exc.resets_in.total_seconds(),
+            )
+
+        self._hit_ratelimits()
+
+        response = self.default_response
+        form = response["activestate_publisher_form"]
+
+        if not form.validate():
+            self.request.session.flash(
+                self.request._("The trusted publisher could not be registered"),
+                queue="error",
+            )
+            return response
+
+        # Check for an already registered publisher before creating.
+        publisher = (
+            self.request.db.query(ActiveStatePublisher)
+            .filter(
+                ActiveStatePublisher.organization == form.organization.data,
+                ActiveStatePublisher.activestate_project_name == form.project.data,
+                ActiveStatePublisher.actor_id == form.actor_id,
+            )
+            .one_or_none()
+        )
+        if publisher is None:
+            publisher = ActiveStatePublisher(
+                organization=form.organization.data,
+                activestate_project_name=form.project.data,
+                actor=form.actor.data,
+                actor_id=form.actor_id,
+            )
+
+            self.request.db.add(publisher)
+
+        # Each project has a unique set of OIDC publishers; the same
+        # publisher can't be registered to the project more than once.
+        if publisher in self.project.oidc_publishers:
+            self.request.session.flash(
+                self.request._(
+                    f"{publisher} is already registered with {self.project.name}"
+                ),
+                queue="error",
+            )
+            return response
+
+        for user in self.project.users:
+            send_trusted_publisher_added_email(
+                self.request,
+                user,
+                project_name=self.project.name,
+                publisher=publisher,
+            )
+
+        self.project.oidc_publishers.append(publisher)
+
+        self.project.record_event(
+            tag=EventTag.Project.OIDCPublisherAdded,
+            request=self.request,
+            additional={
+                "publisher": publisher.publisher_name,
+                "id": str(publisher.id),
+                "specifier": str(publisher),
+                "url": publisher.publisher_url(),
+                "submitted_by": self.request.user.username,
+            },
+        )
+
+        self.request.session.flash(
+            f"Added {publisher} "
+            + (f"in {publisher.publisher_url()}" if publisher.publisher_url() else "")
+            + f" to {self.project.name}",
```

Comment:
I think `publisher_url()` can never be falsey here, right? Based on:

https://github.com/pypi/warehouse/blob/30c8720951db492dbd800bb5433edbdf1fc6085f/warehouse/oidc/models/activestate.py#L125-L128

So, I think this can be:

```suggestion
            f"Added {publisher} "
            + f"in {publisher.publisher_url()} "
            + f"to {self.project.name}",
```

(and then maybe reduce the lines?)

Reply:
Oh man, good eye.

Reply:
Having a bit of trouble with this change.  It looks like the [Google Publisher Model](https://github.com/ActiveState/warehouse/blob/main/warehouse/oidc/models/google.py#L84) returns None for `publisher_url` and I'm not, if that can even be changed, what it would be.

Mind if I leave this change out as well?

Reply:
Actually, this points out an issue in my code.  In my attempted to be type safe, i'm breaking the interface we're using for the PublisherMixin.  I'm going to change `publisher_url` in the AtiveState Publisher mixin to not have the type declarations I added.  This can be iterated on later and done across all published to make sure they match.

Reply:
OK, sorry for the noise.  I DID make this change but I also backed out my additional type declarations that make the `ActiveStatePublisherMixin.publisher_url` not match what the others `*PublisherMixin.publisher_url` have.